spanner/spanner_snippets/spanner/integration_test.go

```diff
@@ -386,6 +388,27 @@ func runBackupSample(t *testing.T, f backupSampleFunc, dbName, backupID, errMsg
 	return b.String()
 }
 
+func runBackupSampleWithRetry(t *testing.T, f backupSampleFunc, dbName, backupID, errMsg string, maxAttempts int) string {
+	var b bytes.Buffer
+	var testError error
+	success := testutil.Retry(t, maxAttempts, time.Minute, func(r *testutil.R) {
+		if err := f(&b, dbName, backupID); err != nil {
+			if spanner.ErrCode(err) == codes.InvalidArgument && strings.Contains(err.Error(), "Please retry the operation once the pending restores complete") {
+				r.Fail()
+			} else {
+				testError = err
+			}
+		}
+	})
+	if !success {
+		t.Errorf("%s: %v", errMsg, "backup test failed because it exceeded max attempts")
+	}
+	if testError != nil {
+		t.Errorf("%s: %v", errMsg, testError)
+	}
```

Comment:
No need to do this. You can use `r.Errorf` instead of `r.Fail` whenever an error happens that you want to fail the retry attempt. If you get an error on every attempt, the test will fail with the right message.

Reply:
@tbpg 
Thanks, that seems reasonable. But what is missing here AFICT is a feature that enables us to break the retry loop when the function returns an error that we don't want to retry. In this specific case only certain errors should be retried. Other errors should preferably fail the test immediately. I've added a feature for that to the test retryer for that. Would you mind taking a look?